src/geopackage/wkb.cpp

```diff
@@ -0,0 +1,306 @@
+    }
+
+    for (uint32_t i = 0; i < count - 1; i++) {
+        auto inner = read_linestring(buffer, index, order);
+        polygon.inners().at(i).reserve(inner.size());
+        for (auto& p : inner) {
+            polygon.inners().at(i).push_back(p);
+        }
+    }
+
+    return polygon;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::multipoint_t wkb::read_multipoint(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    multipoint_t mp;
+    mp.resize(count);
+
+    for (auto& point : mp) {
+        const byte_t new_order = buffer[index];
+        index++;
+
+        uint32_t type;
+        utils::copy_from(buffer, index, type, new_order);
+
+        point = read_point(buffer, index, new_order);
+    }
+
+    return mp;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::multilinestring_t wkb::read_multilinestring(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    multilinestring_t ml;
+    ml.resize(count);
+    for (auto& line : ml) {
+        const byte_t new_order = buffer[index];
+        index++;
+
+        uint32_t type;
+        utils::copy_from(buffer, index, type, new_order);
+
+        line = read_linestring(buffer, index, new_order);
+    }
+
+    return ml;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::multipolygon_t wkb::read_multipolygon(const byte_vector& buffer, int& index, uint8_t order)
+{
+    uint32_t count;
+    utils::copy_from(buffer, index, count, order);
+
+    multipolygon_t mpl;
+    mpl.resize(count);
+    for (auto& polygon : mpl) {
+        const byte_t new_order = buffer[index];
+        index++;
+
+        uint32_t type;
+        utils::copy_from(buffer, index, type, new_order);
+
+        polygon = read_polygon(buffer, index, new_order);
+    }
+
+    return mpl;
+}
+
+// ----------------------------------------------------------------------------
+
+typename wkb::geometry wkb::read(const byte_vector& buffer)
+{
+    if (buffer.size() < 5) {
+        throw std::runtime_error("buffer reached end before encountering WKB");
+    }
+
+    int index = 0;
+    const byte_t order = buffer[index];
+    index++;
+
+    uint32_t type;
+    utils::copy_from(buffer, index, type, order);
+
+    geometry g;
+    switch(type) {
+        case 1:  g = read_point(buffer, index, order); break;
+        case 2:  g = read_linestring(buffer, index, order); break;
+        case 3:  g = read_polygon(buffer, index, order); break;
+        case 4:  g = read_multipoint(buffer, index, order); break;
+        case 5:  g = read_multilinestring(buffer, index, order); break;
+        case 6:  g = read_multipolygon(buffer, index, order); break;
+        default: g = point_t{std::nan("0"), std::nan("0")}; break;
```

Comment:
Should the `default` case also raise an exception, rather than returning an invalid value?